### DISCUSSION TITLE
Tutorial for Collision Avoidance

### DISCUSSION THREAD

**Discussion by devbf** [bitcraze/discussions#733] [Q&A] [answered]
Hello,

We are a small team of students using crazyflies for a small project. At the moment we are using Crazyswarm to use collision avoidance for multiple drones. In our case we have a small swarm of 4 drones which should be able to avoid a collision if another drone flies into the same swarm (while all drones are into the same channel).

We noticed that one of the developers, @jpreiss, has also implemented collision avoidance in the native crazyflie-firmware (https://github.com/bitcraze/crazyflie-firmware/pull/628). But at the moment we are not really sure how we can utilize this function. Using the crazyflie-firmware with cflib commands, we already had some crashes, so it seems that the collision avoidance between two drones is not enabled by default.

Is there any way to use the collision avoidance in the crazyflie-firmware in a way like you can with Crazyswarm using the [enableCollisionAvoidance](https://crazyswarm.readthedocs.io/en/latest/api.html#pycrazyswarm.crazyflie.Crazyflie.enableCollisionAvoidance) function using the cflib?

As the student project currently only uses python programs in combination with cflib and is not really fond of the thought of using Crazyswarm in addition to the existing programs, we would be really glad if someone has a hint on how to use collision avoidance for normal drone flights using the cflib.

Thanks for your answers (moved here from issues because I was told it fits into discussions better)

Flo

**Comment by krichardsson** [comment, marked as answer]
Hi,
There is an example of how to use the collision avoidance functionality in the demo we used at IROS 2022, see https://www.bitcraze.io/about/events/iros2022/. You can also find links to the source code there.

The short version is:
1. The collision avoidance module needs to know the positions of other Crazyflies, this is handled by the [peer localization module](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/iros-2022/src/modules/interface/peer_localization.h). Positions of other Crazyflies are [pushed in here](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/e8992cdca71e443cc7de6074e195bd9390398fec/examples/demos/decentralized_swarm/src/common_files/ds_p2p_interface.c#L110)
2. [Enable collision avoidance](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/e8992cdca71e443cc7de6074e195bd9390398fec/examples/demos/decentralized_swarm/src/pilot.c#LL245C25-L245C25)

In the IROS demo we use the lighthouse system for positioning and peer to peer communication to let the other Crazyflies know about the current positions of Crazyflies. If you have an external positioning system (like a mocap system) and are pushing positions to the Crazyflie using the `packed external pose interface` (as in crazyswarm), [positions of other Crazyflies are automatically sent to the peer localization module](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/e8992cdca71e443cc7de6074e195bd9390398fec/src/modules/src/crtp_localization_service.c#L202-L223)
Unfortunately it does not look like there is an implementation for `packed external pose interface` in the python lib, I think it was added by crazyswarm but it was not implemented in the python lib.

I don't think the collision avoidance module has any functions implemented in the python lib, but it looks like you can enable and control it[ through parameters](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/e8992cdca71e443cc7de6074e195bd9390398fec/src/modules/src/collision_avoidance.c#L390-L428).

**Reply by devbf** [reply]
Hi Kristoffer,

Thank you for your answer and the hint toward those parameters. As they are also in the original firmware ([Link](https://github.com/bitcraze/crazyflie-firmware/blob/c4925370e3c8a35a79bc1e34f9799028751222f6/src/modules/src/collision_avoidance.c#LL390C1-L429C1)), I will try out those before testing the experimental version from IROS.
